system-contracts/test/L1Messenger.spec.ts

```diff
@@ -0,0 +1,531 @@
+      method: "hardhat_stopImpersonatingAccount",
+      params: [TEST_BOOTLOADER_FORMAL_ADDRESS],
+    });
+  });
+
+  describe("publishPubdataAndClearState", async () => {
+    it("publishPubdataAndClearState passes correctly", async () => {
+      await (await l1Messenger.connect(l1MessengerAccount).sendL2ToL1Log(isService, key, value)).wait();
+      numberOfLogs++;
+      await (await l1Messenger.connect(l1MessengerAccount).sendToL1(message)).wait();
+      numberOfMessages++;
+      numberOfLogs++;
+      await (
+        await l1Messenger
+          .connect(knownCodeStorageAccount)
+          .requestBytecodeL1Publication(bytecodeHash, { gasLimit: 130000000 })
+      ).wait();
+      numberOfBytecodes++;
+      // Prepare data for publishPubdataAndClearState()
+      const numberOfLogsBytes = ethers.utils.hexZeroPad(ethers.utils.hexlify(numberOfLogs), 4);
+      const numberOfMessagesBytes = ethers.utils.hexZeroPad(ethers.utils.hexlify(numberOfMessages), 4);
+      const numberOfBytecodesBytes = ethers.utils.hexZeroPad(ethers.utils.hexlify(numberOfBytecodes), 4);
+      // Prepare totalL2ToL1PubdataAndStateDiffs
+      const totalL2ToL1PubdataAndStateDiffs = ethers.utils.concat([
+        numberOfLogsBytes,
+        firstLog,
+        secondLog,
+        numberOfMessagesBytes,
+        currentMessageLengthBytes,
+        message,
+        numberOfBytecodesBytes,
+        lengthOfBytecodeBytes,
+        bytecode,
+        version,
+        compressedStateDiffsSizeBytes,
+        enumerationIndexSizeBytes,
+        compressedStateDiffs,
+        numberOfStateDiffsBytes,
+        encodedStateDiffs,
+      ]);
+      // publishPubdataAndClearState()
+      await (
+        await l1Messenger
+          .connect(bootloaderAccount)
+          .publishPubdataAndClearState(totalL2ToL1PubdataAndStateDiffs, { gasLimit: 10000000 })
+      ).wait();
+      numberOfLogs = 0;
+      numberOfMessages = 0;
+      numberOfBytecodes = 0;
+    });
+
+    it("should revert Too many L2->L1 logs", async () => {
+      await (await l1Messenger.connect(l1MessengerAccount).sendL2ToL1Log(isService, key, value)).wait();
+      numberOfLogs++;
+      await (await l1Messenger.connect(l1MessengerAccount).sendToL1(message)).wait();
+      numberOfMessages++;
+      numberOfLogs++;
+      await (
+        await l1Messenger
+          .connect(knownCodeStorageAccount)
+          .requestBytecodeL1Publication(bytecodeHash, { gasLimit: 130000000 })
+      ).wait();
+      numberOfBytecodes++;
+      // Prepare data for publishPubdataAndClearState()
+      const numberOfMessagesBytes = ethers.utils.hexZeroPad(ethers.utils.hexlify(numberOfMessages), 4);
+      const numberOfBytecodesBytes = ethers.utils.hexZeroPad(ethers.utils.hexlify(numberOfBytecodes), 4);
+      // Prepare totalL2ToL1PubdataAndStateDiffs
+      // set numberOfLogsBytes to 0x900 to trigger the revert (max value is 0x800)
+      const totalL2ToL1PubdataAndStateDiffs = ethers.utils.concat([
```

Comment:
did we need to send logs/messages to make this test pass?

it seems like the only key part of this test is the `0x900`, so all the previous steps are not needed

Reply:
To pass this test successfully, you don't need to follow the previous steps. However, it's important to do so for the next tests that check how revert test cases behave. That's why, in my approach, I completed the whole process of sending logs/messages in the 'should revert Too many L2->L1 logs' test case, instead of breaking it into multiple tests.

In latest changes broken into multiple tests.